### DISCUSSION TITLE
pass reference to another UserData from the index of a UserData

### DISCUSSION THREAD

**Discussion by Absolucy** [mlua-rs/mlua#50] [Q&A] [answered]
Ok so, I have one `UserData` which is this:
```rust
pub struct HttpRequest {
	request: Request<Body>,
	headers: Headers,
}
```
`Headers` is another `UserData` type. I want to override the index metamethod for `httpRequest`, so that `headers` is a field that can be referenced and changed from lua, like `request.headers["X-Powered-By"] = "mlua"`. How could I pass back a reference to the userdata from a metamethod?

**Comment by khvzak** [comment, marked as answer]
If I correctly understand question, you cannot simply return reference to an internal field from a metamethod, unfortunately. 
The easiest way is to wrap `Headers` to `Arc<Mutex< Headers >>` and return a copy instead.

For example:
```rust
use mlua::{chunk, Lua, MetaMethod, Result, UserData, UserDataFields, UserDataMethods};

#[derive(Clone)]
struct Headers(Arc<Mutex<HashMap<String, String>>>);

pub struct HttpRequest {
    request: Request<Body>,
    headers: Headers,
}

impl UserData for HttpRequest {
    fn add_fields<'lua, F: UserDataFields<'lua, Self>>(fields: &mut F) {
        fields.add_field_method_get("headers", |_, this| Ok(this.headers.clone()));
    }
}

impl UserData for Headers {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_meta_method(MetaMethod::Index, |_, this, name: String| {
            Ok(this.0.lock().unwrap().get(&name).cloned())
        });

        methods.add_meta_method_mut(
            MetaMethod::NewIndex,
            |_, this, (name, value): (String, String)| {
                this.0.lock().unwrap().insert(name, value);
                Ok(())
            },
        );
    }
}
```
(valid for mlua `0.6.0-beta.2`)

**Reply by kpfleming** [reply]
Have there been any changes since 0.6.0-beta2 which would provide an alternative method to implement this? I'm in pretty much the same situation, a UserData field is a HashMap, and I want to provide a way to index into that HashMap (for both read and write) with the simplest possible Lua syntax.
